src/library/scala/collection/immutable/HashMap.scala

```diff
@@ -654,6 +879,39 @@ private final class BitmapIndexedMapNode[K, +V](
     new BitmapIndexedMapNode[K, V](dataMap ^ bitpos, nodeMap, dst, dstHashes, size - 1, cachedJavaKeySetHashCode - keyHash)
   }
 
+  /** Variant of `copyAndMigrateFromInlineToNode` which mutates `this` rather than returning a new node.
+    *
+    * @param bitpos the bit position of the data to migrate to node
+    * @param keyHash the improved hash of the key currently at `bitpos`
+    * @param node the node to place at `bitpos`
+    */
+  def migrateFromInlineToNodeInPlace[V1 >: V](bitpos: Int, keyHash: Int, node: MapNode[K, V1]): this.type = {
+    val dataIx = dataIndex(bitpos)
+    val idxOld = TupleLength * dataIx
+    val idxNew = this.content.length - TupleLength - nodeIndex(bitpos)
+
+    val src = this.content
+    val dst = new Array[Any](src.length - TupleLength + 1)
```

Comment:
This array is strictly smaller; are we sure we even need to copy it at all?  Can't we just shift elements in place?

Reply:
Wow that is a very interesting idea...

Reply:
Well, I implemented it, but it broke lots of code in the Champ Common code like ChampBaseIterator, etc. I think this is a good idea and salvageable in long term but maybe out of scope of this PR.

Reply:
The issue is that what would need to happen is that all the data nodes after the removed data need to shift left by 2, and all the nodes before the new node spot need to be shifted left by 1. That leaves a gap of 1 between the last data node and the first node node.  
  
That is fine theoretically but it does require some code changes in for example the `equals` code and `ChampBaseIterator` and likely a few other places.

Reply:
Ok I implemented this feature fully in a separate branch. Repeating what I sent in chat to @Ichoran for visibility and posterity:

I implemented your suggestion fully regarding reusing CHAMP content arrays, just shifting elements around when migrating from inline data to node data inplace.
The benchmarks didn't show any cpu time improvement.
I also went one step further and, when inserting a new data node (key-value), we first check if the content array is large enough (has enough left-over empty slots) to fit the extra key-val pair. Turns out that just slows everything way down
That is, when inserting key-value pairs into this.content, and we're able to mutate, we first check if, as a result of multiple previous data-to-node migrations there has resulted in enough unused space in the array to fit the new key-value
basically
```scala
val minimumContentSize = this.payloadArity * TupleLength + this.nodeArity + TupleLength // minimum size to fit all this node's data + 1 tuple
if (this.content.length >= minimumContentSize) {
  shift things around
} else {
  this.content = ...create new content array...
}
```

Reply:
Well actually, I say it didn't improve things, but that's not true. Here are the benchmarks:

```
// migrate from inline to node inplace creates new content array
[info] # Run complete. Total time: 00:06:51
[info] Benchmark                    (listSize)  (mapSize)  Mode  Cnt         Score         Error  Units
[info] HashMapBenchmarkJosh.concat          10         10  avgt    8       330.043 ±      10.031  ns/op
[info] HashMapBenchmarkJosh.concat          10        100  avgt    8       480.291 ±       7.806  ns/op
[info] HashMapBenchmarkJosh.concat          10       1000  avgt    8       619.731 ±      12.164  ns/op
[info] HashMapBenchmarkJosh.concat          10      10000  avgt    8       956.481 ±      27.692  ns/op
[info] HashMapBenchmarkJosh.concat         100         10  avgt    8      7293.160 ±      73.823  ns/op
[info] HashMapBenchmarkJosh.concat         100        100  avgt    8      5237.407 ±      36.214  ns/op
[info] HashMapBenchmarkJosh.concat         100       1000  avgt    8      7310.880 ±     107.390  ns/op
[info] HashMapBenchmarkJosh.concat         100      10000  avgt    8      8033.495 ±      95.986  ns/op
[info] HashMapBenchmarkJosh.concat        1000         10  avgt    8    111570.167 ±     964.823  ns/op
[info] HashMapBenchmarkJosh.concat        1000        100  avgt    8    117315.137 ±   40411.116  ns/op
[info] HashMapBenchmarkJosh.concat        1000       1000  avgt    8     82535.849 ±     996.116  ns/op
[info] HashMapBenchmarkJosh.concat        1000      10000  avgt    8     97038.737 ±    1056.521  ns/op
[info] HashMapBenchmarkJosh.concat       10000         10  avgt    8   1273457.167 ±    9027.289  ns/op
[info] HashMapBenchmarkJosh.concat       10000        100  avgt    8   1291242.118 ±   13888.651  ns/op
[info] HashMapBenchmarkJosh.concat       10000       1000  avgt    8   1244122.064 ±   10628.704  ns/op
[info] HashMapBenchmarkJosh.concat       10000      10000  avgt    8   1119044.790 ±   61669.128  ns/op


// migrate from inline to node inplace reuses content array

[info] Benchmark                    (listSize)  (mapSize)  Mode  Cnt        Score        Error  Units
[info] HashMapBenchmarkJosh.concat          10         10  avgt    8      325.417 ±     11.622  ns/op
[info] HashMapBenchmarkJosh.concat          10        100  avgt    8      491.547 ±     19.014  ns/op
[info] HashMapBenchmarkJosh.concat          10       1000  avgt    8      638.207 ±     47.864  ns/op
[info] HashMapBenchmarkJosh.concat          10      10000  avgt    8      939.218 ±      9.530  ns/op
[info] HashMapBenchmarkJosh.concat         100         10  avgt    8     6690.524 ±    120.793  ns/op
[info] HashMapBenchmarkJosh.concat         100        100  avgt    8     6094.006 ±    378.198  ns/op
[info] HashMapBenchmarkJosh.concat         100       1000  avgt    8     6918.749 ±    303.648  ns/op
[info] HashMapBenchmarkJosh.concat         100      10000  avgt    8     7963.418 ±    163.125  ns/op
[info] HashMapBenchmarkJosh.concat        1000         10  avgt    8   104492.783 ±    790.520  ns/op
[info] HashMapBenchmarkJosh.concat        1000        100  avgt    8    99903.595 ±    402.805  ns/op
[info] HashMapBenchmarkJosh.concat        1000       1000  avgt    8    77599.793 ±    681.396  ns/op
[info] HashMapBenchmarkJosh.concat        1000      10000  avgt    8    98913.337 ±    790.509  ns/op
[info] HashMapBenchmarkJosh.concat       10000         10  avgt    8  1224895.002 ±   6682.261  ns/op
[info] HashMapBenchmarkJosh.concat       10000        100  avgt    8  1311402.966 ±  14761.165  ns/op
[info] HashMapBenchmarkJosh.concat       10000       1000  avgt    8  1318077.351 ±  17979.071  ns/op
[info] HashMapBenchmarkJosh.concat       10000      10000  avgt    8  1302722.116 ± 181998.502  ns/op
```

I am not sure if we think that that amount of improvement is worth it, but I suppose for Lists around size 100, the gain is pretty significant.

Reply:
I'd just go with whichever code is simpler; it's not that big (or consistent) of a difference.

Reply:
:+1: